[WIP] Add/validation for gamma delta

nitrajka opened the pull request.

nitrajka committed:
Add tests for substitution

ZoltanOnody commented on tests/TestFormula.elm:
```diff
@@ -202,6 +203,65 @@ parseSignedTests =
+original11 = Formula.parseSigned "T \\forall x P(x, k)" |> Result.withDefault (T FF)
+
+validateGamma =
+    describe "gamma substitution test"
```
http://package.elm-lang.org/packages/elm-community/elm-test/4.2.0/Expect#true

ZoltanOnody reviewed (dismissed):
Tiez sa mi nepaci, ze veci co testujes mas definovane inde ako v samotnych testoch, treba potom hore dole behat po subore.

nitrajka committed:
Fix code, so now all the tests pass

nitrajka committed:
Generate documentation and add commands to readme.md

yoyko commented on src/Formula.elm:
```diff
@@ -223,9 +362,29 @@ mapResult f =
 
 
+removeQuantifierAndSubstitute : Substitution -> Formula -> Result String Formula
+removeQuantifierAndSubstitute substitution original =
```
Nie som si uplne isty, ci toto patri do tohoto modulu, a ci by to nemalo byt az v tvojom tablovaci (robi/mala by robit dost vela specifickych veci, vid nizsie|). Tento modul by mal zverejnovat hlavne `substitute` (well, `substFormula` momentalne) funkciu.

Asi je ale mozne, ze sa najdu aj dalsie vyuzitia, takze asi moze byt aj tu, len treba poriadne definovat co vlastne ma robit, momentalne podla toho co robi je to take divne... podla mna by mala robit nasledovne:
- ak je to `∀ x ....` alebo `∃ x ....`  ak `x` je  jedina premenna v substitucii, tak vrati substituovane [1]
  - ak `x` nie je v subst. (alebo su tam aj ine), tak vrati chybu
- inac vrati chybu (lebo nie je quantifier)

[1] - otazka je, co presne robit ak substitucia obsahuje viac premennych... ak to chces pouzit na validaciu v tablovaci, tak by to malo bud zahodit tie ostatne premenne zo substitucie, alebo vratit chybu, ak tomu das substituciu s viac premennymi (preferujem to druhe).

@ZoltanOnody hodi sa nieco take do tvojej prace? co presne by to malo robit?

nitrajka replied:
No takze este raz. Napisem tu viac case-ov. 
Testovat mam `substitute` (po premenovani) alebo tu moju `removeQuantifierAndSubstitute`? Lebo ak mam testovat obe, tak toto vsetko budem musiet definovat pre obe.:

1. Ak moja formula vyzera `∀ x ....` alebo `∃ x ....` a:
    * `x` je v substitucii
        * a je tam viac premennych -> toto sa mi v tableau stat nemoze, ale vo Formule ano a tam to treba testovat. Tu mi napada, ze ta metoda `removeQuantifierAndSubstitute` by mala robit redukciu. Teda nejakym sposobom konverotvat vstup z mojho tableau na vstup pre Formulovsky vseobecny `substitute` a zaroven potom konvertovat jej vystup aby boli pravidla substitucie v tableau dodrzane. Vies o com hovorim?
            Plati to co som nakodila? Ze budem odstranovat kvantifikatory s premennymi kym sa ta premenna nachadza v substitucii?
        Cize ak je tam viac premennych tak v tableau testoch (na `removeQuantifierAndSubstitute`) vyhodim chybu ale vo formule to nevadi. Lenze moze sa mi vobec stat ze z tableau poslem v substitucii viac premennych? Nemoze. Takze ma zmysel to testovat?
        * nie je tam viac premennych -> Ok, substituujem a vratim to.
    * `x` nie je v substitucii -> V tableau je problem a vo formule nie, tak? V testoch pre formulu (`substitute`) to teda zasubstituujem a vratim. V testoch pre moje tableau (`removeQuantifierAndSubstitute`) mam ocakavat chybu. Tak?
2. ak moja formula nema na zaciatku quantifier. -> Toto je v pohode vo vseobecnej Formule (`substitute`) ale nie v tableau. Takze ta metoda `removeQuantifierAndSubstitue` dava celkom aj zmysel. A este mi dava zmysel aj to aby som testy pre `substitute` nechala vo Formule ale tu metodu `removeQuantifierAndSubstitue` presunula do mojho tableau a aj testy k nej napisala.

Co hovoris? @yoyko Dava to zmysel?

yoyko replied:
Nechaj skludom removeQuantifierAndSubstitute v tomto module... myslim, ze sa to hodi aj Zolimu potom...

Co sa tyka testovania:

Vacsina testov by mala byt pre `substitute`, tu chces otestovat vsetky mozne cases so substituciou:
- normalna substitucia,
-  viacero premennych (Ze je to naraz),
- ze viazane premenne to nesubstitije (tj. napriklad x->a do "P(x)&\forall x Q(x)" ma spravit "P(a)&\forall x Q(x)" a pod)
- ze to vrati chybu ked subst. nie je aplikovatelna


`removeQuant...AndSubst...` potom staci otestovat na tie "specialne" pripady, ktora ona riesi:
- vrati chybu ak formula nie je "\forall ? ...." alebo "\exists ? ...",
- vrati chybu ak substitucia nie je prave na tu jednu premennu co je vo vonk. kvant.
- viac menej staci otestovat iba jeden/dva kladne pripady (exists a forall?), hlavne ze korektne urve ten kvantifikator...

nitrajka replied:
Oki, dakujem 😃

nitrajka replied:
@yoyko , prave som si uvedomia, ze Janciho (@crnkjck ) kod nevracia chybu ked substitucia nie je aplikovatelna. Uz som jancimu aj pisala mail, ale zatial nic.

Napr: `\\forall n \\forall k P(n,k)` a `n->k, k->n` mi nevrati chybu, ale povodnu nesubstituovanu formulu. On si totiz tu:
https://github.com/FMFI-UK-1-AIN-412/elm-formula/pull/2/files#diff-c892343f4203d314590b2465177e1ac6R301 
vsetko nesubstituovatelne (viazane) pomaze zo substitucie, cize tuto:
https://github.com/FMFI-UK-1-AIN-412/elm-formula/pull/2/files#diff-c892343f4203d314590b2465177e1ac6R273 
sa potom nevykona `canSubst`. Lebo v lokalnej premennej `σ` uz nie je to co na zaciatku. a zas keby to odtial nemazem, tak mi zasubstituuje aj viazane...

yoyko commented on src/Formula.elm:
```diff
@@ -223,9 +362,29 @@ mapResult f =
+            if List.member s (Dict.keys substitution) then
+                removeQuantifierAndSubstitute substitution f
+            else
+                substFormula substitution f |> Result.map (ForAll s)
```
Tu to teda asi chce chybu ...
(A plust teda mozno celkovo aj ked `substitution` ma viac premennych)

yoyko commented on src/Formula.elm:
```diff
@@ -223,9 +362,29 @@ mapResult f =
+                substFormula substitution f |> Result.map (Exists s)
+
+        _ ->
+            substFormula substitution original
```
A tu tiez

yoyko commented on src/Formula.elm:
```diff
@@ -97,6 +104,115 @@ type alias Substitution =
+
+
+substitute : Substitution -> Formula -> Int -> Formula
+substitute substitution original depth =
```
Toto vsetko v tejto casti teda vyzera ako nepouzite veci?
(Myslim, ze sme sa rozpravali, ze pouzime tu Janciho verziu?)

nitrajka replied:
Ano, toto je outdated, pouzila som Janciho verziu.

yoyko commented on src/Formula.elm:
```diff
@@ -150,6 +266,25 @@ freeTerm t =
+
+
+allVariablesInFormula : Formula -> Set String
+allVariablesInFormula f =
```
Rozmyslam, ci na nieco treba zoznam vsetkych premennych (v porovnani s vsetkymi volnymi premennymi, co je nizsie). Asi to ma zmysel pre upnost ;)

Nazval by som to mozno `variablesInFormula` a to `freeFormula` nizsie na `freeVariablesInFormula` ? (A potom ten `freeTerm` hore, co je asi `variablesInTerm`).

Druha alternativa je `formulaVariables`, `termVariables`, `freeFormulaVariables` (tie prve dve sa mi pacia, to tretie trochu skripe)

@crnkjck  mas ty nejake preferencie na nazvy? :)

nitrajka replied:
Nie, netreba, je to outdated. `allVariablesInFormula` je zmazana. Ale `freeFormula` a `freeTerm` mozem premenovat.

yoyko commented on src/Formula.elm:
```diff
@@ -223,9 +362,29 @@ mapResult f =
 substFormula : Substitution -> Formula -> Result String Formula
 substFormula σ f =
     let
+        canSubst : String -> Term -> Set String -> Result String Term
```
Toto `substFormula` by som asi teda premenoval na `substitute`

yoyko commented on src/Formula.elm:
```diff
@@ -280,8 +439,9 @@ substFormula σ f =
-        substTs σ bound =
-            mapResult (substT σ bound)
+        substTs : Substitution -> Set String -> List Term -> Result String (List Term)
+        substTs σ bound lst =
```
Mne sa viac pacil ten predchadzajuci point free styl ;)

nitrajka replied:
Yoyo, nehnevaj sa, ja to tam celkom aj potrebujem. 😀Je mi zrejmejsie co to robi.

yoyko commented on src/Formula.elm:
```diff
@@ -644,6 +821,7 @@ binary conn constructor =
             |. symbol ")"
 
 
+quantified : List String -> (String -> Formula -> b) -> Parser b
```
`b` by tu asi explicitne mohlo byt `Formula` ;)

nitrajka replied:
Oki

yoyko commented on src/Formula.elm:
```diff
@@ -713,6 +893,18 @@ spaces =
+strSubstitution s =
+    "("
+        ++ Dict.foldl
+            (\key -> \value -> \meh -> key ++ "->" ++ strTerm value ++ ", ")
```
Tu ignorujes/zahadzujes predch. vysledok foldu, takze to nepospajas, ale vysledoku bude iba posledna "premenna->term,".
Plus to dava `,` aj za poslednym, co tak (samozrejme asi so zatvorkami naokolo)
```elm
s
  |> Dict.toList
  |> List.map (\(v,t) -> v ++ "->" ++ strTerm t)
  |> String.join ","
```

A pridaj si unit testy (s viac premennymi v substitucii ;)

nitrajka replied:
Oki

yoyko commented on tests/TestFormula.elm:
```diff
@@ -202,6 +203,91 @@ parseSignedTests =
+testSubstitution s original new =
+    test
+        ("in formula " ++ (original |> Formula.strFormula) ++" using subs " ++ (Formula.strSubstitution s))
+        (\() -> Expect.equal (Ok new) (Formula.removeQuantifierAndSubstitute s original))
```
Tieto testy by mali testovat  `substFormula` (reps `substitute` ked to premenujes).
Samozrejme mozes spravit aj nejake testy na to specialne spravanie `removeQuanti.....`

nitrajka replied:
Oki

yoyko commented on tests/TestFormula.elm:
```diff
@@ -202,6 +203,91 @@ parseSignedTests =
+            (Formula.parse "\\forall x \\forall k P(x,k)" |> Result.withDefault FF)
+        , testSubstitution
+            (Dict.fromList [("x", Var "y"), ("y", Var "x")])
+            (Atom "P" [Var "x", Var "y"]) -- P(x,y)
```
Btw, hore su definovane vselijake helpre pre `p`, `x`, `y` a pod... (akurat pozor, ze niektore z nich robia viacpismenkove mena, napr "yyy").
Bolo by to s nimi trochu kratsie/citalnejsie (mozno)
```elm
p [x, y]
p [ f [ x ] ]
```
a pod.

nitrajka replied:
Oki

yoyko commented:
Grr, prepac, pozeram, ze som si myslel, ze pozeram all changes, ale mal som otvoreny iba prvy commit...
Ignoruj komenty ktore su k starym veciam ;) (ale niektore z tych outdated su stale platne, github ich tak ukazuje aj ked si potom napr len presunula na ine miesto alebo tak)

Este narychlo pozriem tie ostatne veci

nitrajka committed:
Change removeQuantifierAndSubstitute to throw errors

I also refactored tests. Now I test (already) renamed
function `substitute` and `removeQuantifierAndSubstitute` separately.

ZoltanOnody commented:
Nechceme to pomaly mergnut? Idealne do bakalarky by som dal odkaz uz na verziu ktora je na elm packages.

nitrajka commented:
Chceme, prosim mergni to. Ja tam uz nemam co dodat.
